webrtc/RTCDataChannel-send.html

```diff
@@ -0,0 +1,380 @@
+  //  assertEqualsArrayBuffer
+
+  /*
+    6.2.  RTCDataChannel
+      interface RTCDataChannel : EventTarget {
+        ...
+        readonly  attribute RTCDataChannelState readyState;
+        readonly  attribute unsigned long       bufferedAmount;
+                  attribute EventHandler        onopen;
+                  attribute EventHandler        onmessage;
+                  attribute DOMString           binaryType;
+
+        void send(USVString data);
+        void send(Blob data);
+        void send(ArrayBuffer data);
+        void send(ArrayBufferView data);
+      };
+
+      bufferedAmount
+        The bufferedAmount attribute must return the number of bytes of application
+        data (UTF-8 text and binary data) that have been queued using send() but that,
+        as of the last time the event loop started executing a task, had not yet been
+        transmitted to the network. (This thus includes any text sent during the
+        execution of the current task, regardless of whether the user agent is able
+        to transmit text asynchronously with script execution.) This does not include
+        framing overhead incurred by the protocol, or buffering done by the operating
+        system or network hardware. If the channel is closed, this attribute's value
+        will only increase with each call to the send() method (the attribute does not
+        reset to zero once the channel closes).
+
+
+    [WebMessaging]
+    interface MessageEvent : Event {
+      readonly attribute any data;
+      ...
+    };
+   */
+
+  // Simple ASCII encoded string
+  const helloString = 'hello';
+  // ASCII encoded buffer representation of the string
+  const helloBuffer = Uint8Array.of(0x68, 0x65, 0x6c, 0x6c, 0x6f);
+  const helloBlob = new Blob([helloBuffer]);
+
+  // Unicode string with multiple code units
+  const unicodeString = '世界你好';
+  // UTF-8 encoded buffer representation of the string
+  const unicodeBuffer = Uint8Array.of(
+    0xe4, 0xb8, 0x96, 0xe7, 0x95, 0x8c,
+    0xe4, 0xbd, 0xa0, 0xe5, 0xa5, 0xbd);
+
+  /*
+      6.2.  send()
+        2.  If channel's readyState attribute is connecting, throw an InvalidStateError.
+
+   */
+  test(t => {
+    const pc = new RTCPeerConnection();
+    const channel = pc.createDataChannel('test');
+    assert_equals(channel.readyState, 'connecting');
+    assert_throws('InvalidStateError', () => channel.send(helloString));
+  }, 'Calling send() when data channel is in connecting state should throw InvalidStateError');
+
+  /*
+      6.2.  send()
+        4.  If channel's underlying data transport is not established yet, or if the
+            closing procedure has started, then abort these steps.
+   */
+  test(t => {
+    const pc = new RTCPeerConnection();
+    const channel = pc.createDataChannel('test');
+    channel.close();
+    assert_equals(channel.readyState, 'closing');
+    channel.send(helloString);
+  }, 'Calling send() when data channel is in closing state should succeed');
```

Comment:
> If channel's underlying data transport is not established yet, or if the closing procedure has started, then abort these steps.

My understanding for this is that if `readyState` is not `connected`, the method is aborted without actually sending the data. 

The case for _"channel's underlying data transport is not established yet"_ is a bit curious, because step 2 have already guard against connecting state.

I think the reason it is abort in close state is because in the current version, any error sending would abruptly close the connection and the exception is swallowed.

I have raised some comment in w3c/webrtc-pc#1209 that if error sending is changed to throwing, then it should throw error if connection is closed as well. Perhaps this test can be deferred until that is resolved.